src/main/kotlin/uk/gov/justice/digital/hmpps/externalusersapi/service/UserService.kt

```diff
@@ -0,0 +1,70 @@
+package uk.gov.justice.digital.hmpps.externalusersapi.service
+
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.reactor.awaitSingleOrNull
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.externalusersapi.jpa.repository.GroupRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.jpa.repository.RoleRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.jpa.repository.UserRepository
+import uk.gov.justice.digital.hmpps.externalusersapi.r2dbc.data.User
+import uk.gov.justice.digital.hmpps.externalusersapi.security.AuthSource
+import java.util.Optional
+
+@Service
+@Transactional(readOnly = true)
+class UserService(
+  private val roleRepository: RoleRepository,
+  private val userRepository: UserRepository,
+  private val groupRepository: GroupRepository,
+) {
+  @Transactional
+  suspend fun getUser(username: String?): User? {
+
+    var user = userRepository.findByUsernameAndSource(username).awaitSingleOrNull()
+    Optional.of(user!!).orElseThrow()
+    val groups = user.id?.let {
+      groupRepository.findGroupsByUserId(it)
+    }?.toList()
+
+    val roles = user.id?.let { roleRepository.findRoleByUserId(it) }?.toList()
```

Comment:
If we have a user then we should have an id as it's the primary key - wondering if we can improve this rather than having user.id?

Reply:
NO ACTION - previously discussed